[FLINK-1819][core]Allow access to RuntimeContext from Input and Output formats

sachingoel0101 opened the pull request:
1. Introduces new Rich Input and Output formats, similar to Rich Functions.
2. Makes all existing input and output formats rich, without any API breaking changes.
3. Provides RuntimeContext to GenericSink and GenericSource operators
4. DataSourceTask and DataSinkTask set DistributedUDFContext to their input and output formats respectively.
5. Complete access to RuntimeContext, including access to accumulators.

mxm commented:
Thanks for the pull request!

You did not convert all InputFormats and OutputFormats. There are a few more, e.g. HadoopInputFormatBase or JDBCInputFormat. Other than that, your pull request looks good. I think it is fine to keep the notion of normal and rich, just like we do for the operator user functions. That also means that we do not break the existing interfaces.

mxm commented on flink-runtime/src/main/java/org/apache/flink/runtime/operators/DataSourceTask.java:
```diff
@@ -75,6 +78,13 @@
+	/**
+	 * The accumulator map used in the RuntimeContext.
+	 */
+	protected Map<String, Accumulator<?,?>> accumulatorMap;
```
This field should be final. Or you get rid of the field and just access the accumulators via `getEnvironment().getAccumulatorRegistry().getUserMap()`.

sachingoel0101 commented:
Ah yes. I'll update them in a while. There's actually some problem with the unit test I've written too. Travis fails sporadically.

sachingoel0101 commented:
@mxm , I've ported all existing IO formats to Rich IO formats. The unit test is consistently passing now.

I also added a `postConfigure` function which allows user to access runtime context and setup things which use the runtime context, for example, accumulators. Its usage can be found in the unit test I've added. I found it impossible to guarantee the existence of `RuntimeContext` in the `configure` function itself.

StephanEwen commented:
Would not not be much simpler to add an interface `RuntimeContextFormat` with a method (`setRuntimeContext()`) and call this method on the formats that implement this interface? That would be three changed classes, done.

I am also a bit sceptical about the difference between input format and rich input format. For transformation functions the difference between Function (a lambda) and RichFunction (many mode methods) makes sense. Input formats are per-se pretty rich...

StephanEwen commented:
What is the `postConfigure` method for? Input and output formats have already the `open()` method, that is called after `configure()`. The `configure()` method itself is somewhat of a baggage from the past, when all code was shipped as classes + configuration, rather than via closures.

sachingoel0101 commented:
@StephanEwen, it would be much simpler to have an interface. But then we leave the part about implementing the `setRuntimeContext` and `getRuntimeContext` to the user. IMO that's not a very good option. So it makes to extend rich formats from abstract Rich formats. Doing that, I couldn't figure out a better solution if we want to make all existing IO formats rich. To sum it up:
1. If we only defined an interface, we leave the implementation of `setter` and `getter` methods to the user.
2. If we write a Rich abstract class, we have to make all existing IO formats rich since the user might want to extend one of these, and of course can't extend two classes.
If there's a way to achieve this, let me know. I might have missed something.

As for the `postConfigure` method, it was to guarantee that the `RuntimeContext` will be available now. I tried initializing the context in `DataSourceTask` and `DataSinkTask` before calling `configure`, but it seems the format is re-initialized somewhere as that information is lost. I could remove this function and we can just initialize `counters` etc. in `open` call.

fhueske commented:
Hi @sachingoel0101,

thanks for opening this PR! 
I see your point that users would have to implement `setRuntimeContext()` and `getRuntimeContext()` if we would use an interface. However, I am not sure that this would affect many people because we could implement the methods in all our base Input/OutputFormats which should cover a large portion of new formats. Also, these methods can be implemented in less than a minute. I think @StephanEwen has a point that all Input/OutputFormats are always quite rich. I would opt for the interface solution but implement it for all relevant Input/OutputFormats that Flink provides.

I would also remove the `postConfigure()` method in favor of a clear documentation of the interface stating that the `RuntimeContext`is available in `open()`.

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/recordJobs/util/DiscardingOutputFormat.java:
```diff
@@ -19,14 +19,14 @@
  * A simple output format that discards all data by doing nothing.
  */
-public class DiscardingOutputFormat implements OutputFormat<Record> {
+public class DiscardingOutputFormat extends RichOutputFormat<Record> {
```
I don't think the DiscardingOutputFormat needs to access the RuntimeContext

fhueske commented on flink-core/src/test/java/org/apache/flink/api/common/operators/util/NonRichGenericInputFormat.java:
```diff
@@ -0,0 +1,86 @@
+ * This is copied from {@link org.apache.flink.api.common.io.GenericInputFormat}
+ * This class doesn't provide access to RuntimeContext.
+ */
+public abstract class NonRichGenericInputFormat<OT> implements InputFormat<OT, GenericInputSplit> {
```
I don't think we need this class. Why would somebody need a generic input format without RuntimeContext, if you can have one with RC for free?

sachingoel0101 replied:
This was only for testing purposes. I needed to verify that I didn't break any functionality by making every IO format rich. That is, to verify the class casting behavior in the Pact tasks.

fhueske replied:
Ah, sorry. Didn't notice it is in test scope. Good have it for tests!

sachingoel0101 replied:
No problem! :)

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/io/DiscardingOutputFormat.java:
```diff
@@ -27,7 +27,7 @@
  * @param <T> The type of the elements accepted by the output format.
  */
-public class DiscardingOutputFormat<T> implements OutputFormat<T> {
+public class DiscardingOutputFormat<T> extends RichOutputFormat<T> {
```
Doesn't need to access the RuntimeContext.

fhueske commented on flink-java/src/test/java/org/apache/flink/api/java/type/extractor/TypeExtractorInputFormatsTest.java:
```diff
@@ -130,7 +131,7 @@ public void testQueryableFormatType() {
 	// --------------------------------------------------------------------------------------------
 	
-	public static final class DummyFloatInputFormat implements InputFormat<Float, InputSplit> {
+	public static final class DummyFloatInputFormat extends RichInputFormat<Float, InputSplit> {
```
Test input/output formats don't need access to the RuntimeContext.

fhueske commented on flink-staging/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseWriteStreamExample.java:
```diff
@@ -80,7 +80,7 @@ public void cancel() {
 	 *
 	 */
-	private static class HBaseOutputFormat implements OutputFormat<String> {
+	private static class HBaseOutputFormat extends RichOutputFormat<String> {
```
No need to update this class if the example doesn't use the RuntimeContext.

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/accumulators/AccumulatorLiveITCase.java:
```diff
@@ -307,7 +307,7 @@ public void flatMap(String value, Collector<Integer> out) throws Exception {
 	}
 
-	private static class WaitingOutputFormat implements OutputFormat<Integer> {
+	private static class WaitingOutputFormat extends RichOutputFormat<Integer> {
```
Test OF does not use the RuntimeContext.

sachingoel0101 commented:
@fhueske , what do we settle on then? Should I keep it as abstract class or change it to an interface?
Also, which classes need to be Rich? Can I assume that the ones you didn't comment on are okay?

sachingoel0101 commented:
@fhueske, in the life cycle of an Input format, the `open` method is called several times in each parallel instance of the `DataSource` operator. Consider for example, the simple example of adding an accumulator. Since access to `RuntimeContext` is enabled only after configure, we will end up adding the same accumulator several times, which of course would fail. This is why I decided to add a `postConfigure` method.

StephanEwen commented:
I am pretty much against `postConfigure()`. There are so many life cycle methods already, we should really not make it more complicated (I think even `configure()` is a problem, but that's an artifact).

It is not so hard to make it right in the `open()` method, and there are methods that getOrCreate accumulators, afaik.

sachingoel0101 commented:
Okay I guess we could leave the proper usage of `RuntimeContext` to the user. We make it explicit that the `RuntimeContext` will not be available in `configure` in the documentation, and if someone is overriding the `open` method, they likely know it'll be called several times and can take care of not performing, for example, duplicate `accumulator` additions. What do you think?

sachingoel0101 commented:
Build failure is on unrelated tests. The recent error on `PartitionedStateCheckpointingITCase` and the other on `KafkaITCase`.

fhueske commented on flink-core/src/main/java/org/apache/flink/api/common/operators/GenericDataSinkBase.java:
```diff
@@ -328,9 +330,12 @@ public int compare(IN o1, IN o2) {
 		}
-		
+
+		if(format instanceof RichOutputFormat){
```
Move after `configure()` call

fhueske commented on flink-core/src/main/java/org/apache/flink/api/common/operators/GenericDataSourceBase.java:
```diff
@@ -200,14 +202,18 @@ public void accept(Visitor<Operator<?>> visitor) {
 		@SuppressWarnings("unchecked")
 		InputFormat<OUT, InputSplit> inputFormat = (InputFormat<OUT, InputSplit>) this.formatWrapper.getUserCodeObject();
+
+		if(inputFormat instanceof RichInputFormat){
```
Move after `configure()` call

fhueske commented on flink-core/src/test/java/org/apache/flink/api/common/io/RichInputFormatTest.java:
```diff
@@ -0,0 +1,45 @@
+
+	@Test
+	public void testCheckRuntimeContextAccess() {
+		final SerializedInputFormat<Record> inputFormat = new SerializedInputFormat<Record>();
```
We are trying to reduce the use of the `Record` data type which is an artifact of a deprecated API. Could you replace it by another data type?

fhueske commented on flink-core/src/test/java/org/apache/flink/api/common/io/RichOutputFormatTest.java:
```diff
@@ -0,0 +1,45 @@
+
+	@Test
+	public void testCheckRuntimeContextAccess() {
+		final SerializedOutputFormat<Record> inputFormat = new SerializedOutputFormat<Record>();
```
Please replace `Record` data type.

fhueske commented on flink-staging/flink-jdbc/src/main/java/org/apache/flink/api/java/record/io/jdbc/JDBCOutputFormat.java:
```diff
@@ -40,7 +40,7 @@
 import org.apache.flink.types.StringValue;
 import org.apache.flink.types.Value;
 
-public class JDBCOutputFormat implements OutputFormat<Record> {
```
Class has been removed by a recent commit.

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/classloading/jar/CustomInputSplitProgram.java:
```diff
@@ -65,7 +65,7 @@ public static void main(String[] args) throws Exception {
 	// --------------------------------------------------------------------------------------------
 	
-	public static final class CustomInputFormat implements InputFormat<Integer, CustomInputSplit>, ResultTypeQueryable<Integer> {
+	public static final class CustomInputFormat extends RichInputFormat<Integer, CustomInputSplit> implements ResultTypeQueryable<Integer> {
```
This is a test class which does not need to be adapted.

fhueske commented on flink-staging/flink-hadoop-compatibility/src/main/java/org/apache/flink/hadoopcompatibility/mapred/record/HadoopRecordInputFormat.java:
```diff
@@ -37,7 +37,7 @@
 import org.apache.hadoop.util.ReflectionUtils;
 
-public class HadoopRecordInputFormat<K, V> implements InputFormat<Record, HadoopInputSplit> {
+public class HadoopRecordInputFormat<K, V> extends RichInputFormat<Record, HadoopInputSplit> {
```
Class will be removed soon. Not strictly necessary to adapt it.

fhueske commented on flink-staging/flink-hadoop-compatibility/src/main/java/org/apache/flink/hadoopcompatibility/mapred/record/HadoopRecordOutputFormat.java:
```diff
@@ -37,7 +37,7 @@
 
 
-public class HadoopRecordOutputFormat<K,V> implements OutputFormat<Record> {
+public class HadoopRecordOutputFormat<K,V> extends RichOutputFormat<Record> {
```
Class will be removed soon. Not strictly necessary to adapt it.

fhueske commented on flink-core/src/main/java/org/apache/flink/api/common/io/ReplicatingInputFormat.java:
```diff
@@ -59,7 +60,7 @@
  * @see org.apache.flink.api.common.operators.base.MapPartitionOperatorBase
  */
-public final class ReplicatingInputFormat<OT, S extends InputSplit> implements InputFormat<OT, S> {
+public final class ReplicatingInputFormat<OT, S extends InputSplit> extends RichInputFormat<OT, S> {
```
The `ReplicatingInputFormat` can wrap any kind of `InputFormat` (with or without RuntimeContext). 
Therefore, it needs to extend `RichInputFormat` but overwrite the `setRuntimeContext()` method and forward the context to its wrapped IF, if this is also a `RichInputFormat`.

Please check the updated JavaDocs as well. The wrapped IF does not need to be rich.

fhueske commented:
Hi @sachingoel0101,
Thanks for the update! The PR looks good in general.
Regarding the interface vs. abstract class issue, I am a bit undecided with a slight preference towards the abstract class. However, I agree with @StephanEwen that `RichInputFormat` and `RichOutputFormat` are not good names. How about `ContextInputFormat` and `ContextOutputFormat`?

Any other opinions on that matter?

sachingoel0101 commented:
I'm inclined towards the Abstract class too. If we're saying something is context aware, we better provide access to it IMO. It doesn't make sense to ask the user to write it themselves and guarantee that we'd just call the relevant methods at runtime.
As for the names, wouldn't it be better to maintain consistency in naming? All the * rich * functions are named `RichFunctions`, which is why I decided to name these `RichFormat`s too. I do agree though. The IO formats are already pretty * rich * in the sense that they allow the user to control everything that happens at runtime.

mxm commented:
@StephanEwen @fhueske 
The `Rich` prefix might seem a bit odd but it is a naming convention that is consistent with the user-defined functions which have the RuntimeContext available.
As for `abstract classes vs interface` I agree with @sachingoel0101 that it makes sense to implement access to the RuntimeContext once for the user. We may also add other "rich" functions; if we implemented that using an interface we would have to change it later on or add another interface.

I think Input/Ouput formats should be "rich" by default. However, we might not want to break the API for existing external implementations.

StephanEwen commented:
Okay, so how about calling the abstract base class`AbstractInputFormat` and have it implement the runtime context and leaves the other methods abstract?

sachingoel0101 commented:
Okay. I think maybe calling them `Rich` maybe an overkill. I will change the names to `Abstract`. 
Do we keep the existing formats *non-rich* or would it be okay to make all of them *rich* [which is the case right now]? The problem with not making them *rich* is that then we're limiting any extending classes to never be *rich*.

fhueske commented:
I'm in favor of making the existing formats "rich".

mxm commented:
Yes, existing formats should converted to "rich" formats. The name `Abstract..` makes more sense if it becomes the new default way of interfacing with the rich input/output formats. I still think, calling it `Rich...` makes more sense in terms of consistency with the user-defined functions.

sachingoel0101 commented:
Okay. I'll hold off on making any changes right now. We should reach at a consensus first.

fhueske commented:
@mxm, @sachingoel0101 get your point. Being consistent with the remaining API is a valid argument. I would be OK with calling the abstract classes `RichInputFormat` and `RichOutputFormat`.

Let's wait for at least 24 hours and merge this PR if nobody speaks up.

StephanEwen commented:
We are back to square one ;-)

  - `Function`: single abstract method
  - `RichFunction` = 5 methods. I see how that gets rich.

  - `InputFormat`: 8 methods
  - `RichInputFormat`: 10 methods.
We could call it `SlightlyMoreRichInputFormat` ;-) I cannot help but find this very confusing.

Why the urgency to stick with the "Rich" prefix?

StephanEwen commented:
For transformation functions, there is a clear case for "thin" versus "rich", for Java8 lambdas.
Input formats are a different game. They are super rich by default anyways.

StephanEwen commented:
The suggestion `AbstractInputFormat` was not so bad, in my opinion.
If you want a name that explains what's happening, you can always call it `InputFormatWithContext` ;-)

fhueske commented:
To keep it consistent with the remaining API. For functions you need to extend a RichFunction if you want to have access to the RuntimeContext. I agree that the name is not perfect (and I think everybody else got your point as well) but I think it's a valid point to aim for consistency.

mxm commented:
"Rich" does not refer to the number of methods but the fact that it has the RuntimeContext available. All non-rich variants do not get state inserted. This follows a naming convention in Flink. `AbstractInputFormat` might be a more intuitive name for novices but I'm more inclined to naming consistency.

StephanEwen commented:
Functions also need to extend RichFunction to have access to `open()` and `close()`.
I think the two things a different enough that any strife for "consistency" is actually pretty "random".
If your thoughts currently revolve around the RuntimeContext, it apprears more consistent. If you thoughts are on the life cycle methods, it seems inconsistent. Random.

I think you should go ahead and just call them "Rich". It is just a name, and what matters is that the JavaDocs describe what it actually means...

sachingoel0101 commented:
I agree with Stephan's argument that addition of context to I/O formats is
a very marginal enhancement. He literally stole my words. :')
However, from my perspective, when I first started using flink, Rich meant
runtime context. The idea of open and close wasn't as nearly exciting as
the runtime context.
What if we changed back to the original name mentioned on jira and make it
`ContextAwareInputFormat`? Would everyone be okay with that?
On Aug 12, 2015 8:10 PM, "Stephan Ewen" <notifications@github.com> wrote:

> Functions also need to extend RichFunction to have access to open() and
> close().
> I think the two things a different enough that any strife for
> "consistency" is actually pretty "random".
> If your thoughts currently revolve around the RuntimeContext, it apprears
> more consistent. If you thoughts are on the life cycle methods, it seems
> inconsistent. Random.
>
> I think you should go ahead and just call them "Rich". It is just a name,
> and what matters is that the JavaDocs describe what it actually means...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/966#issuecomment-130324909>.
>

sachingoel0101 commented:
Oh apologies. I  only saw the first comment on the email thread. I guess
it's more or less settled. I'll leave it up to you guys to make a final
decision on this. :')
On Aug 12, 2015 10:59 PM, "Sachin Goel" <sachingoel0101@gmail.com> wrote:

> I agree with Stephan's argument that addition of context to I/O formats is
> a very marginal enhancement. He literally stole my words. :')
> However, from my perspective, when I first started using flink, Rich meant
> runtime context. The idea of open and close wasn't as nearly exciting as
> the runtime context.
> What if we changed back to the original name mentioned on jira and make it
> `ContextAwareInputFormat`? Would everyone be okay with that?
> On Aug 12, 2015 8:10 PM, "Stephan Ewen" <notifications@github.com> wrote:
>
>> Functions also need to extend RichFunction to have access to open() and
>> close().
>> I think the two things a different enough that any strife for
>> "consistency" is actually pretty "random".
>> If your thoughts currently revolve around the RuntimeContext, it apprears
>> more consistent. If you thoughts are on the life cycle methods, it seems
>> inconsistent. Random.
>>
>> I think you should go ahead and just call them "Rich". It is just a name,
>> and what matters is that the JavaDocs describe what it actually means...
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/apache/flink/pull/966#issuecomment-130324909>.
>>
>

mxm commented:
Thank you for your remarks @StephanEwen.

>I think you should go ahead and just call them "Rich". It is just a name, and what matters is that the JavaDocs describe what it actually means...

The JavaDocs in this pull request properly describe what a "RichInputFormat" is. So we are good to go. Let's stick with "Rich" although I understand your concerns.

I'll merge this later if nobody objects in the meantime :)

mxm commented:
Thanks for the contribution @sachingoel0101!

sachingoel0101 commented:
Sure. :)
